ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+   * @param name the unique name for the store under compaction.
+   * @return whether compaction is in progress for the store.
+   */
+  static boolean isCompactionInProgress(String dir, String name) {
+    return new File(dir, name + COMPACTION_LOG_SUFFIX).exists();
+  }
+
+  /**
+   * Creates a new compaction log.
+   * @param dir the directory at which the compaction log must be created.
+   * @param name unique name of the store.
+   * @param time the {@link Time} instance to use.
+   * @param compactionDetailsList the details about the compaction.
+   */
+  CompactionLog(String dir, String name, Time time, List<CompactionDetails> compactionDetailsList) throws IOException {
+    this.time = time;
+    file = new File(dir, name + COMPACTION_LOG_SUFFIX);
+    if (file.exists() || !file.createNewFile()) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + " already exists or could not be created");
+    }
+    startTime = time.milliseconds();
+    cycleLogs = new ArrayList<>(compactionDetailsList.size());
+    for (CompactionDetails details : compactionDetailsList) {
+      cycleLogs.add(new CycleLog(details));
+    }
+    flush();
+  }
+
+  /**
+   * Loads an existing compaction log.
+   * @param dir the directory at which the log exists.
+   * @param name unique name of the store.
+   * @param storeKeyFactory the {@link StoreKeyFactory} that is used for keys in the {@link BlobStore} being compacted.
+   * @param time the {@link Time} instance to use.
+   */
+  CompactionLog(String dir, String name, StoreKeyFactory storeKeyFactory, Time time) throws IOException {
+    this.time = time;
+    file = new File(dir, name + COMPACTION_LOG_SUFFIX);
+    if (!file.exists()) {
+      throw new IllegalArgumentException(file.getAbsolutePath() + " does not exist");
+    }
+    try (FileInputStream fileInputStream = new FileInputStream(file)) {
+      CrcInputStream crcInputStream = new CrcInputStream(fileInputStream);
+      DataInputStream stream = new DataInputStream(crcInputStream);
+      short version = stream.readShort();
+      switch (version) {
+        case VERSION_0:
+          startTime = stream.readLong();
+          currentIdx = stream.readInt();
+          int cycleLogsSize = stream.readInt();
+          cycleLogs = new ArrayList<>(cycleLogsSize);
+          while (cycleLogs.size() < cycleLogsSize) {
+            cycleLogs.add(CycleLog.fromBytes(stream, storeKeyFactory));
+          }
+          long crc = crcInputStream.getValue();
+          if (crc != stream.readLong()) {
+            throw new IllegalStateException("CRC of data read does not match CRC in file");
+          }
+          break;
+        default:
+          throw new IllegalArgumentException("Unrecognized version");
+      }
+    }
+  }
+
+  /**
+   * @return the current state of compaction.
+   */
+  State getCompactionState() {
+    return currentIdx >= cycleLogs.size() ? State.DONE : getCurrentCycleLog().getState();
```

Comment:
Yeah - it's just a standard check. The > should not happen. I don't think it affects correctness so would prefer to let it be.